### PR TITLE
Add options to highlight pkm on map

### PR DESCRIPTION
**Pull request by michikrug**
## Description
This PR adds options and functionality to configure color highlighting to "good" pkm.
The user can configure to colors for "High IV" pkm (default: red), "High Level" pkm (default: green) and "Perfect" pkm (default: purple).
There is also an option to add a circle symbol with "100" to "Perfect" pkm.

I have added thee different ways of processing:
1) `highlight-pokemon: css` will apply the outlining with CSS drop-shadow - the user can configure color and blur radius
2) `highlight-pokemon: svg` will apply the outlining with SVG and shadow markers - the user can configure color and blur radius as well as the size

In addition, an option in the frontend is added to scale the pkm markers by their values (level, iv). Also, the zIndex is modified to move better pkm to the front.

Furthermore, also the default rendering of generated pkm markers has been improved by adding a bit of black shadow around them. This increases the contrast and visibility in my opinion.

## Motivation and Context
To improve the visibility of "good" pkm a visual clue (color highlight) is added.

## How Has This Been Tested?
Own instance.

## Screenshots (if appropriate):
Bildschirmfoto 2022-03-11 um 00 10 42
Screenshot_20220310-161144

## Types of changes
- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist:
- [x] My code follows the code style of this project.
- [x] My change requires a change to the documentation.
- [x] I have updated the documentation accordingly.

### CODE REVIEW

**Comment by michikrug** [comment]
Btw. as I was unsure what options may be useful to be customized, I would be open and happy for feedback in terms of changes or additions.

**Comment by hamster007Github** [comment]
Tested successful. Great work :thumbsup:  Hope it will be merged.
Possible improvements I have in mind: 
1. high IV and high lvl thresholds configurable
2. any chance to add this as options in the user configuration menu instead of config.ini? So every user can choose themself.

**Comment by Hupe7** [comment]
works great for me. can be released

**Comment by michikrug** [comment]
As requested I will add configuration options to the frontend

**Comment by michikrug** [comment]
Currently experimenting with moving the highlighting completely to the client side.
We would have different options here:

1) CSS Drop Shadow:
Screenshot_20220312-170308

2) SVG Circle Marker Shadow:
Screenshot_20220312-170302

3) Current server-side approach (circle can still be kept for the others, I guess):
Screenshot_20220312-170313


WDYT?

**Comment by michikrug** [comment]
For the people who are interested in trying the client side options: give my pr-highlight2 branch a try
https://github.com/michikrug/RocketMAD/tree/pr-highlight2

The configuration options changed a bit.
I included all three proposed options that can be selected using the `highlight-pokemon` config option.
You can choose between `server` (the initial solution), `css` (css drop shadow) or `svg` (a circular shadow using svg)

**Comment by wraythex** [comment]
pr-highlight2 is working great here!  I really like the customization options being available per user.  I'm using the SVG option and it is working great.

I don't know if this is a Safari/iPhone issue, but in Safari the background circles are solid instead of fading away towards the edges.  Not really a problem for me, but wanted to let you know.

If you can, could the color options be a drop down list, or a grid of colors to choose from?
And, is there a way to highlight 0% IV pokemon?

Thanks again for this great qol pr!

**Comment by michikrug** [comment]
> pr-highlight2 is working great here! I really like the customization options being available per user. I'm using the SVG option and it is working great.
> 
> I don't know if this is a Safari/iPhone issue, but in Safari the background circles are solid instead of fading away towards the edges. Not really a problem for me, but wanted to let you know.
> 
> If you can, could the color options be a drop down list, or a grid of colors to choose from? And, is there a way to highlight 0% IV pokemon?
> 
> Thanks again for this great qol pr!

Thanks for testing.

1) yes, this seems to be a Safari issue. The blurring can somehow not be applied. I will test some other ways...

2) drop down list would be kind of limiting (or what colors should we add?) a color picker would be cool, I will see what I can do

3) currently, there is no 0% highlight option available. would probably work the same way with an additional switch

**Comment by michikrug** [comment]
Merged pr-highlighting2 branch into this one.

**Comment by michikrug** [comment]
Added color picker

**Comment by wraythex** [comment]
Thanks for making the changes!  However, when I switched to this PR just now, pokemon that are highlighted do not show up on the map, only the highlight color shows up.  
missing

Hovering the mouse over the highlight color did show the pokemon in the pop-up, just not on the map.

Pokemon with no highlight color showed up normally.

I did not see any errors in the browsers console.  The only difference I see between pr-highlight2 and pr246 for the missing pokemon is in the classes applied to the div:
pr-highlight2 (showing pokemon): `<div class="leaflet-marker-icon leaflet-zoom-animated leaflet-interactive"`
pr246 (not showing pokemon): `<div class="leaflet-marker-icon marker-highlight leaflet-zoom-animated leaflet-interactive"`

Please let me know if I can try anything or provide any other information to help track this issue down.

**Comment by michikrug** [comment]
Oh yes. I did a minor mistake with the classes / css. Now both the shadow as the pkm is blurred...

Will fix this tomorrow. Thanks.

**Comment by wraythex** [comment]
This latest update is great!  Mons are showing up in the color circles, color picker is working great, and color circles look correct in Safari!  Thanks again for these great additions!

**Comment by hamster007Github** [comment]
Sorry for my late response. Tested last version of PR and it is greate 👍 . Community loves it. Go for the svg version for now. Is there any difference according browser performance impact or compatibility between svg and css version?
@daangroot would be greate if you can have a look and merge this PR into master.

File: templates/map.html [hunk 1]
```diff
@@ -662,6 +691,74 @@ <h4>{{ i18n('Export Pokémon lists') }}</h4>
...
+                  </div>
+                  {% endif %}
+                  <div class="form-control">
+                    <span id="highlight-iv-slider-title" class="slider-title">min. IVs</span>
+                    <div class="slider-container">
+                      <div id="highlight-iv-slider"></div>
+                    </div>
+                  </div>
+                  <div class="form-control">
+                    <span id="highlight-level-slider-title" class="slider-title">min. Level</span>
```
Comment (na-ji) [inline templates/map.html:746]: Could you please use i18n on those labels too?

**Review by na-ji — APPROVED** [review]
Very good PR. Personally, I would try to not create another generated icon suffix, since it would add more load on the server side, but we're already doing it, so we can ignore this until a future refactor. 
The “CSS” version and “SVG” are equivalent when it comes to performance, but the CSS drop shadow allows having a shadow taking the same shape as the icon. I will merge it when the i18n is sorted out. 
Thanks for this great feature. 😉

File: templates/map.html (hunk 1 above)
Comment (michikrug) [reply templates/map.html:746]: Sure...

**Comment by michikrug** [comment]
> Very good PR. Personally, I would try to not create another generated icon suffix, since it would add more load on the server side, but we're already doing it, so we can ignore this until a future refactor. The “CSS” version and “SVG” are equivalent when it comes to performance, but the CSS drop shadow allows having a shadow taking the same shape as the icon. I will merge it when the i18n is sorted out. Thanks for this great feature. 😉

Would you say we should remove the `server` variant and only keeping the 100 circle setting? Or is this still a valid option to keep?

**Comment by na-ji** [comment]
> Would you say we should remove the `server` variant and only keeping the 100 circle setting? Or is this still a valid option to keep?

Yes, I believe the other two options are more than enough.

**Comment by michikrug** [comment]
I removed the server variant and added i18n for the labels.

_Please note the changed config parameters!_

**Review by na-ji — APPROVED** [review]

**Comment by na-ji** [comment]
Thank you @michikrug 👍

**Pull request merged**